### ISSUE TITLE
Potential deadlock when executing engine jobs

### ISSUE THREAD

**Issue by hubertp** [enso-org/enso#7941]
It appears that a user can still occasionally reach a state when engine becomes unresponsive and all (visualization) requests timeout.

Here is a log collected by @jdunkerley 
[enso-project-manager-2023-09-29.0.zip](https://github.com/enso-org/enso/files/12781740/enso-project-manager-2023-09-29.0.zip)
during one of such sessions.
There is nothing special in it. `DetachVisualization` requests is followed by `ApplyEdit` ones and eventually the handler timeouts. 

What is probably missing in that log is that `DetachVisualizationJob` is submitted but never appears to be executed, hence the timeout.
Additionally, when one wants to shutdown LS completely it throws:
```
[ERROR] [2023-09-29T13:45:42+01:00] [enso] Error executing ExecuteJob(contextId=627db9f1-e478-4509-883b-6f64469e6b74)
java.lang.InterruptedException: null
        at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireInterruptibly(AbstractQueuedSynchronizer.java:959)
        at java.base/java.util.concurrent.locks.ReentrantLock$Sync.lockInterruptibly(ReentrantLock.java:161)
        at java.base/java.util.concurrent.locks.ReentrantLock.lockInterruptibly(ReentrantLock.java:372)
        at org.enso.interpreter.instrument.execution.ReentrantLocking.logLockAcquisition(ReentrantLocking.scala:123)
        at org.enso.interpreter.instrument.execution.ReentrantLocking.acquireContextLock(ReentrantLocking.scala:106)
        at org.enso.interpreter.instrument.job.ExecuteJob.run(ExecuteJob.scala:33)
        at org.enso.interpreter.instrument.job.ExecuteJob.run(ExecuteJob.scala:18)
        at org.enso.interpreter.instrument.execution.JobExecutionEngine.$anonfun$runInternal$1(JobExecutionEngine.scala:113)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
        at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread.access$001(PolyglotThread.java:53)
        at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$1.execute(PolyglotThread.java:100)
        at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$ThreadSpawnRootNode.executeImpl(PolyglotThread.java:134)
        at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$ThreadSpawnRootNode.execute(PolyglotThread.java:125)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:718)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:641)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callBoundary(OptimizedCallTarget.java:574)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.doInvoke(OptimizedCallTarget.java:558)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callIndirect(OptimizedCallTarget.java:486)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.call(OptimizedCallTarget.java:467)
        at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread.run(PolyglotThread.java:96)
```

I initially thought it's a fluke but no longer thinks so. It manifests an important problem with the execution engine - a number of jobs are being queued (including `DetachVisualizationJob` or `ExecuteJob`) and all of them appear to be stuck. 

We need to discover what sequence of locks leads to this unfortunate scenario.

**Comment by hubertp** [comment]
I think `ExecuteJob` acquires all the necessary context locks, starts to execute a job but somehow never finishes. At least that's how I understand lack of message saying that a context lock has been released. Would need a complete thread dump to confirm.

**Comment by hubertp** [comment]
This is kind of related to https://github.com/enso-org/enso/issues/7059.
So the scenario here is that `Attach/DetachVisualizationCmd` will acquire `context` lock before doing any work, similarly `Attach/DetachVisualizationJob`, which is triggered by it. But the command can be blocked, potentially for a long time, by a long running background `ExecuteJob` because it also requires `context` lock while it runs. I think commands should never even block or we will constantly deal with the timeouts headache.

**Comment by hubertp** [comment]
Fixed via #7953
